chain-time/src/timeframe.rs

```diff
@@ -0,0 +1,168 @@
+    pub slot: Slot,
+    /// The offset of a specific time frame in
+    pub slot_duration_offset: Duration,
+}
+
+/// Time frame which is a timeline that is configured to be split in discrete slots
+pub struct TimeFrame {
+    timeline: Timeline,
+    slot_offset: Slot,
+    slot_duration: NbSeconds,
+    //slot_configuration: Vec<u64>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct NbSeconds(u64);
+
+impl NbSeconds {
+    pub fn new(seconds: u32) -> Self {
+        assert!(seconds < 600);
+        NbSeconds(seconds as u64)
+    }
+
+    pub fn to_duration(&self) -> Duration {
+        Duration::from_secs(self.0)
+    }
+}
+
+impl TimeFrame {
+    /// Create a new time frame with a specific slot size
+    ///
+    /// ```text
+    ///
+    /// 0        1        2        3        4        5                  
+    /// x--------x--------┳--------x--------x--------x  frame ticking at per_slot
+    ///
+    /// ^
+    /// |
+    /// timeline
+    /// ```
+    ///
+    pub fn new(timeline: Timeline, per_slot: NbSeconds) -> Self {
+        TimeFrame {
+            timeline,
+            slot_offset: Slot(0),
+            slot_duration: per_slot,
+        }
+    }
+
+    /// Change time frame at a specific slot
+    ///
+    /// Note this also change the beginning of this time frame, to start
+    ///
+    /// ```text
+    /// 0        1        2        3        4        5                  
+    /// x--------x--------┳--------x--------x--------x  frame ticking at NbSeconds(9)
+    ///                   |
+    ///                   ┕---x---x---x---x---x         returned frame
+    ///                   2   3   4   5   6   7
+    ///                   ↑
+    ///                   |
+    ///                   frame.change_frame(Slot(2), NbSeconds(4))
+    /// ```
+    ///
+    pub fn change_frame(&self, slot: Slot, duration_per_slot: NbSeconds) -> Self {
+        let d = Duration::from_secs(slot.0 * self.slot_duration.0);
+        let new_timeline = self.timeline.advance(d);
+        TimeFrame {
+            timeline: new_timeline,
+            slot_offset: Slot(self.slot_offset.0 + slot.0),
+            slot_duration: duration_per_slot,
+        }
+    }
+
+    /// Given a system time get the slot and associated duration leftover
+    pub fn to_slot_and_remaining(&self, at: &SystemTime) -> Option<SlotAndDuration> {
+        match self.timeline.differential(at) {
+            None => None,
+            Some(t) => {
+                let slot_nb = t.0.as_secs() / self.slot_duration.0;
+                let e = slot_nb * self.slot_duration.0;
+                let d = t.0 - Duration::from_secs(e); // cannot wrap
+                Some(SlotAndDuration {
+                    slot: Slot(self.slot_offset.0 + slot_nb),
+                    slot_duration_offset: d,
+                })
+            }
+        }
+    }
+
+    /// Given a system time get the slot associated
+    pub fn to_slot(&self, at: &SystemTime) -> Option<Slot> {
```

Comment:
The naming of this and the above method is a bit misleading. `to_` is normally used for one-to-other conversions, while this uses system time and self as the basis.
I think `slot_at` would be a better name.

Reply:
Done